Web API wrapper modules

mthjones opened the pull request:
This is it! All of the Slack web API endpoints have been wrapped. However, there are still a few outstanding issues that **should be resolved before this gets merged** or at the very least least tracked to get resolved after merging.

- [x] ~~No endpoints are tested except for `channels.*`, `api.*` and `auth.*`. [**In Progress**](https://github.com/BenTheElder/slack-rs/pull/17#issuecomment-148960255)~~
- [ ] `files.upload` is left unimplemented. This requires pulling in other dependencies (namely [`multipart`](https://crates.io/crates/multipart)), but the dependencies require some TLC to work with what we have going on.
- [x] ~~Some of the parameters still take `Option<&str>` when an `Option<bool>` (or just `bool`?) or `Option<u32>` or some other more strongly-typed value would be more appropriate. Still trying to figure out how to work with this, because `Option<u32>` in particular does not want to play nice.~~
- [x] ~~Weak documentation~~
- [x] Still a few response values that don't perfectly line up with the actual returned value. ~~In particular, there are some responses that return a list of heterogeneous "`Item`"s, which I'm not sure what to do with yet.~~ There are also some responses that have a field called `type`, which is a keyword and requires a custom decoder like `RtmStart`/`rtm::StartResponse`.
- [x] General cleanup and code consistency pass
- [x] ~~Still have the `mock.rs` module from `hyper`. I want to remove this. Looking into either using [`yup-hyper-mock`](https://crates.io/crates/yup-hyper-mock) or seeing if `hyper` can expose the `mock` module itself.~~

The commits in this PR will get squashed prior to merge. 

**I welcome any feedback/criticism/bikeshedding!**

BenTheElder commented:
Awesome work! I'll review tomorrow afternoon (have a midterm at 2 tomorrow)
and try to get this through asap.

About the mocking: Have you talked to the hyper dev(s) about exposing mock?
I think other projects could benefit from that as well. If not we should
look into that.
On Oct 15, 2015 23:40, "Matt Jones" <notifications@github.com> wrote:

> This is it! All of the Slack web API endpoints have been wrapped. However,
> there are still a few outstanding issues that *should be resolved before
> this gets merged* or at the very least least tracked to get resolved
> after merging.
>
>    - No endpoints are tested except for channels.*
>    - files.upload is left unimplemented. This requires pulling in other
>    dependencies (namely multipart <https://crates.io/crates/multipart>),
>    but the dependencies require some TLC to work with what we have going on.
>    - Some of the parameters still take Option<&str> when an Option<bool>
>    or Option<u32> or some other more strongly-typed value would be more
>    appropriate. Still trying to figure out how to work with this, because
>    Option<u32> in particular does not want to play nice.
>    - Weak documentation
>    - Still a few response values that don't perfectly line up with the
>    actual returned value. In particular, there are some responses that return
>    a list of heterogeneous "Item"s, which I'm not sure what to do with
>    yet. There are also some responses that have a field called type,
>    which is a keyword and requires a custom decoder like RtmStart/
>    rtm::StartResponse.
>    - General cleanup and code consistency pass
>    - Still have the mock.rs module from hyper. I want to remove this.
>    Looking into either using yup-hyper-mock
>    <https://crates.io/crates/yup-hyper-mock> or seeing if hyper can
>    expose the mock module itself.
>
> Breaking Changes
>
>    - The existing APIs in RtmClient that wrapped web API endpoints have
>    been removed
>    - The top-level slack namespace no longer exports Group, User, etc.
>    They have been moved under slack::web::api.
>
> The commits in this PR will get squashed prior to merge.
>
> *I welcome any feedback/criticism/bikeshedding!*
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/BenTheElder/slack-rs/pull/17
> Commit Summary
>
>    - Add API scaffold and im.* api
>    - Add api.* methods
>    - Add auth.* methods
>    - Add channels.* API + cleanup
>    - Add chat.* API support
>    - Ran rustfmt on the existing API implementations
>    - Add license to mock module
>    - Add auto-generated files for the rest of the apis to get a head-start
>    - Regenerate the auto-generated apis more properly
>    - Fix up api.test response
>    - Finish emoji.* API
>    - Implement files.* API (mostly)
>    - Implement oauth.* API
>    - Re-add make_authed_api_call and add more general make_api_call
>    - Implement groups.* API
>    - Fix incorrect semicolon
>    - Switch tabs to spaces
>    - Implement users.* API
>    - Fix dead_code annotation
>    - Implement team.* API
>    - Partially implement stars.* API
>    - Partially implement pins.* API
>    - Fix compiler errors and warnings
>    - Implement rtm.* API
>    - Partially implement reactions.* API
>    - Implement search.* API
>    - Integrate web api into lib.rs
>    - Merge branch 'master' of github.com:BenTheElder/slack-rs into
>    base-api-mod
>    - Use new API for channels_history
>    - Remove unused error file
>    - Remove web API wrappers from RtmClient
>    - Move web API structs into web API module
>
> File Changes
>
>    - *M* src/lib.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-0> (606)
>    - *A* src/web/api/api.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-1> (33)
>    - *A* src/web/api/auth.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-2> (25)
>    - *A* src/web/api/channels.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-3> (650)
>    - *A* src/web/api/chat.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-4> (146)
>    - *A* src/web/api/emoji.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-5> (19)
>    - *A* src/web/api/files.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-6> (88)
>    - *A* src/web/api/groups.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-7> (258)
>    - *A* src/web/api/im.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-8> (111)
>    - *A* src/web/api/mock.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-9> (139)
>    - *A* src/web/api/mod.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-10> (239)
>    - *A* src/web/api/oauth.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-11> (27)
>    - *A* src/web/api/pins.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-12> (63)
>    - *A* src/web/api/reactions.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-13> (109)
>    - *A* src/web/api/rtm.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-14> (150)
>    - *A* src/web/api/search.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-15> (107)
>    - *A* src/web/api/stars.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-16> (76)
>    - *A* src/web/api/team.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-17> (73)
>    - *A* src/web/api/users.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-18> (80)
>    - *A* src/web/mod.rs
>    <https://github.com/BenTheElder/slack-rs/pull/17/files#diff-19> (1)
>
> Patch Links:
>
>    - https://github.com/BenTheElder/slack-rs/pull/17.patch
>    - https://github.com/BenTheElder/slack-rs/pull/17.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BenTheElder/slack-rs/pull/17>.
>

BenTheElder commented:
It looks like using `yup-hyper-mock` is the accepted solution: https://github.com/hyperium/hyper/issues/334 (there are no plans to release mock for external usage, so that breaking changes internally are not breaking changes for hyper)

BenTheElder commented:
As per the breaking changes, I think maybe `slack` should re-export `Group`, `User`, etc for convenience.

Still reading through the code, looks great so far, I'd probably want to switch to `yup-hyper-mock` or similar, re-export some of the slack data types (in particular those the RTM client would have direct use for at least), and one or two nits i'll leave as comments on the source.


Also: Not something that needs to be fixed in this PR but; I need to find out the rust convention for indendation (spaces / tabs) and reformat after it goes through, lib.rs (here and master) at least has some weird indentation that aligns in my editor but not online because it's both 4 space tabs and hard tabs.

BenTheElder commented on src/web/api/files.rs:
```diff
@@ -0,0 +1,88 @@
+    pub paging: super::Pagination
+}
+
+/// Uploads or creates a file.
```
Please note in the docs that this is currently unsupported or implement.

BenTheElder commented on src/web/api/mock.rs:
```diff
@@ -0,0 +1,139 @@
+/*
```
We should probably remove this and switch to `yup-hyper-mock` as noted in the PR discussion.

BenTheElder commented:
Overall looks very nice, clean. I'll be happy to merge this and release a new version when it's complete.

Other notes: The custom decoder is unfortunate, but while `rustc-serialize` still gets special treatment on stable, we'll keep it for now, when custom derive is stable i'll replace it with serde which has an accompanying macro for aliasing field names.

That just leaves:
> Some of the parameters still take Option<&str> when an Option<bool> or Option<u32> or some other more strongly-typed value would be more appropriate. Still trying to figure out how to work with this, because Option<u32> in particular does not want to play nice.

And:

>Still a few response values that don't perfectly line up with the actual returned value. In particular, there are some responses that return a list of heterogeneous "Item"s, which I'm not sure what to do with yet. There are also some responses that have a field called type, which is a keyword and requires a custom decoder like RtmStart/rtm::StartResponse.

As well as docs and mocking/testing, I think so far the documentation is pretty reasonable, users should probably refer to the slack api docs. We can add more examples later as there is more real usage.
Mocking / testing, we can work on adding more once you've settled on the mocking to be used, possibly added in a later PR.

I'll look into the other two later this weekend, I'm off for a bit.

BenTheElder commented:
This closes https://github.com/BenTheElder/slack-rs/issues/14 and https://github.com/BenTheElder/slack-rs/issues/7

mthjones commented:
Hope the midterm went well! :)

> Have you talked to the hyper dev(s) about exposing mock?

Not yet. I see you found an issue where the author of `yup-hyper-mock` already raised that issue. I don't agree with the accepted result that there's now a potential disparity between the mocked hyper and the actual hyper since the code is just copied, but we can deal with that for now. I may re-raise the issue with the hyper team – or just make a PR to have hyper use yup-hyper-mock.

> I think maybe slack should re-export Group, User, etc for convenience.

Sure, that's an easy change to make. There's currently no *need* for them to be exported at the top-level though, since with the recent changes clients should never need to directly create them.

> I need to find out the rust convention for indendation (spaces / tabs) and reformat after it goes through, lib.rs (here and master) at least has some weird indentation that aligns in my editor but not online because it's both 4 space tabs and hard tabs.

Heh...always a contentious issue. I've been using spaces in this PR and that's what I typically use in most projects. I don't think Rust has any community style guidelines (apart from `rustfmt`) yet, but FWIW, the [rust stdlib codebase](https://github.com/rust-lang/rust/tree/master/src/libstd) uses spaces.

> Please note in the docs that this is currently unsupported or implement.

Sure. I'd prefer to support it but we could just mark it unsupported for now and implement it later.

> The custom decoder is unfortunate, but while rustc-serialize still gets special treatment on stable, we'll keep it for now, when custom derive is stable i'll replace it with serde which has an accompanying macro for aliasing field names.

Yeah, I'd really love for serde to hit stable. serde does support codegen though, which may be worth looking into as a bridge until serde hits stable. Also has `serde_json`, which could be used as a more explicit JSON converter.

---

Currently working on integrating `yup-hyper-mock` in.

BenTheElder commented:
Thanks! I think it did.

>Not yet. I see you found an issue where the author of yup-hyper-mock already raised that issue. I don't agree with the accepted result because there's now a potential disparity between the mocked hyper and the actual hyper since the code is just copied, but we can deal with that for now. I may re-raise the issue with the hyper team – or just make a PR to have hyper use yup-hyper-mock.

Sounds good. I think `yup-hyper-mock` won't have too hard of a time keeping the file synchronized and that keeps it centralized, but it might be worth re-raising, I think other projects like eg servo would love mocking http requests if they aren't already using `yup-hyper-mock` and it would be better to keep it coupled to the implementation it's mocking I think...

>There's currently no need for them to be exported at the top-level though, since with the recent changes clients should never need to directly create them.

True but if you want to EG store a user in some structure in a bot then `slack::Group` reads better than `slack::web::api::Group` I think, especially in the context of `RtmClient`. Not a big deal either way though.

>Heh...always a contentious issue. I've been using spaces in this PR and that's what I typically use in most projects. I don't think Rust has any community style guidelines (apart from rustfmt) yet, but FWIW, the rust stdlib codebase uses spaces.

4 Spaces is the way to go then. I think writing in go is the only time I've used hard tabs.

>Sure. I'd prefer to support it but we could just mark it unsupported for now and implement it later.

Exactly, either works, I just meant if it doesn't get implemented we should also note it in the docs instead of just the runtime error.

>Yeah, I'd really love for serde to hit stable. serde does support codegen though, which may be worth looking into as a bridge until serde hits stable. Also has serde_json, which could be used as a more explicit JSON converter.

I'll look at the codegen, but for now rustc-serialize seems to be doing the job. I definitely look forward to using the alias macro, golang's struct tags spoiled me :-)

mthjones commented:
Tasks complete:

* Switched to `yup-hyper-mock` and removed `mock.rs`
* Improved documentation, especially around the exported structs
* Implemented the custom decoder for `Item`, which now properly represents it as an enum of `Message | File | File+Comment`. Still issues with several others, especially [`Message`](https://api.slack.com/events/message). I think we can probably do the same thing for it as for `Item`.
* Just commented out `files.upload` for now. Leaving it there but unimplemented felt like a major gotcha for any users.
* Added tests for the `api.*` and `auth.*` endpoints. I really wish there was a way to do this more automated, because it's a huge pain.

BenTheElder commented:
Sounds good.
I don't think there's any good way to automate the tests either unfortunately, but at least once they're written they shouldn't need much updating.

mthjones commented:
Tested modules:

- [x] `api`
- [x] `auth`
- [x] `channels`
- [ ] `chat`
- [x] `emoji`
- [ ] `files`
- [ ] `groups`
- [ ] `im`
- [x] `oauth`
- [x] `pins`
- [ ] `reactions`
- [ ] `rtm`
- [ ] `search`
- [ ] `stars`
- [ ] `team`
- [ ] `users`

Mostly putting this here so I can keep track of what's been completed, but if you're interested I'd love some help hammering the rest of these out. I'm generally taking the sample response data from [the docs](https://api.slack.com/methods) and pasting it in as a raw string to mock responses. There are a few places where the Slack docs are broken though, so it might be useful to use the tester to get some real data, just make sure you scrub the sensitive data! Let me know if there's any you're interested in. I'll be tackling `chat` next.

BenTheElder commented:
None in particular but if I get a chance before you get to it I'll work
from the other end of the list.
On Oct 17, 2015 7:22 PM, "Matt Jones" <notifications@github.com> wrote:

> Tested modules:
>
>    - api
>    - auth
>    - channels
>    - chat
>    - emoji
>    - files
>    - groups
>    - im
>    - oauth
>    - pins
>    - reactions
>    - rtm
>    - search
>    - stars
>    - team
>    - users
>
> Mostly putting this here so I can keep track of what's been completed, but
> if you're interested I'd love some help hammering the rest of these out.
> I'm generally taking the sample response data from the docs
> <https://api.slack.com/methods> and pasting it in as a raw string to mock
> responses. There are a few places where the Slack docs are broken though,
> so it might be useful to use the tester to get some real data, just make
> sure you scrub the sensitive data! Let me know if there's any you're
> interested in. I'll be tackling chat next.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BenTheElder/slack-rs/pull/17#issuecomment-148960255>.
>

mthjones commented:
Whew, tests are complete! Ended up fixing a few bugs that were exposed when writing the tests as well. 

I believe that just leaves implementing the [`Message`](https://api.slack.com/events/message) type as an `enum` and writing a custom decoder for it. Which may be another large chunk of work after looking at the number of message types there are. :weary:

Oh, and doing a quick audit.

mthjones commented:
Hoo boy, this should be fun. The docs for the message events don't actually match the returned data. I'm not sure if `attachments` is supposed to be the `item`, because its fields don't match the items from other endpoints. Man, this documentation has been rough on actually matching the returned data.

See:

[The docs example](https://api.slack.com/events/message/pinned_item):
```json
{
    "type": "message",
    "subtype": "pinned_item",
    "user": "U024BE7LH",
    "item_type": "F",
    "text": "<@U024BE7LH|cal> pinned their Image <https:...7.png|7.png> to this channel.",
    "item": {
        …
    },
    "channel": "C02ELGNBH",
    "ts": "1360782804.083113"
}
```

vs.

A stripped actual message event:
```json
{
    "type": "message",
    "subtype": "pinned_item",
    "user": "U12345678",
    "item_type": "C",
    "attachments": [
        {
            "color": "D0D0D0",
            "fallback": "[October 18th, 2015 11:47 PM] mthjones: ayy lmao",
            "author_subname": "mthjones",
            "ts": "1445226476.000002",
            "text": "ayy lmao",
            "author_link": "test-team.slack.com/team/mthjones",
            "author_icon": "https://secure.gravatar.com/avatar/PRIVATE_GUID.jpg",
            "mrkdwn_in": ["text"]
        }
    ],
    "text": "<@U12345678|mthjones> pinned a message to this channel.", 
    "channel": "C12345678",
    "ts": "1445226479.000003"
}
```

BenTheElder commented:
Yeah, the slack docs are not 100% up to date it seem.
The biggest gotcha so far has been that in the past when i parsed slack
timestamps (in go I think it was) I found out that slack doesn't use a
single timestamp format. (https://github.com/slackhq/slack-api-docs/issues/7
)

I got wrapped up in a project for class, will try to get back to this soon
but I don't know when.

On Sun, Oct 18, 2015 at 11:56 PM, Matt Jones <notifications@github.com>
wrote:

> Hoo boy, this should be fun. The docs for the message events don't
> actually match the returned data. I'm not sure if attachments is supposed
> to be the item, because its fields don't match the items from other
> endpoints. Man, this documentation has been rough on actually matching the
> returned data.
>
> See:
>
> The docs example <https://api.slack.com/events/message/pinned_item>:
>
> {
>     "type": "message",
>     "subtype": "pinned_item",
>     "user": "U024BE7LH",
>     "item_type": "F",
>     "text": "<@U024BE7LH|cal> pinned their Image <https:...7.png|7.png> to this channel.",
>     "item": {
>         …
>     },
>     "channel": "C02ELGNBH",
>     "ts": "1360782804.083113"
> }
>
> vs.
>
> A stripped actual message event:
>
> {
>     "type": "message",
>     "subtype": "pinned_item",
>     "user": "U12345678",
>     "item_type": "C",
>     "attachments": [
>         {
>             "color": "D0D0D0",
>             "fallback": "[October 18th, 2015 11:47 PM] mthjones: ayy lmao",
>             "author_subname": "mthjones",
>             "ts": "1445226476.000002",
>             "text": "ayy lmao",
>             "author_link": "test-team.slack.com/team/mthjones",
>             "author_icon": "https://secure.gravatar.com/avatar/PRIVATE_GUID.jpg",
>             "mrkdwn_in": ["text"]
>         }
>     ],
>     "text": "<@U12345678|mthjones> pinned a message to this channel.",
>     "channel": "C12345678",
>     "ts": "1445226479.000003"
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BenTheElder/slack-rs/pull/17#issuecomment-149091679>.
>

mthjones commented:
No worries. I'm just chipping away at this when I get the chance. :)

I think the *only* thing left now is finishing up the tests for `MessageEvent` and a tidiness pass. *phew*

mthjones commented:
So I forgot to mention, there are a few breaking changes that should be communicated when a new version is released:

- `RtmClient::post_message` now returns `Result<api::chat::PostMessageResponse, Error>`
- `RtmClient::delete_message` now returns `Result<api::chat::DeleteResponse, Error>`
- `RtmClient::mark` now returns `Result<api::channels::MarkResponse, Error>`
- `RtmClient::set_topic` now returns `Result<api::channels::SetTopicResponse, Error>`
- `RtmClient::set_purpose` now returns `Result<api::channels::SetPurposeResponse, Error>`
- `RtmClient::add_reaction_timestamp` now returns `Result<api::reactions::AddResponse, Error>`
- `RtmClient::add_reaction_file` now returns `Result<api::reactions::AddResponse, Error>`
- `RtmClient::add_reaction_file_comment` now returns `Result<api::reactions::AddResponse, Error>`
- `RtmClient::update_message` now returns `Result<api::chat::UpdateResponse, Error>`
- `RtmClient::im_open` now returns `Result<api::im::OpenResponse, Error>`
- `RtmClient::channels_history` now returns `Result<api::channels::HistoryResponse, Error>`
- `RtmClient::im_close` now returns `Result<api::im::CloseResponse, Error>`
- `RtmClient::im_history` now returns `Result<api::im::HistoryResponse, Error>`
- `RtmClient::im_list` now returns `Result<api::im::ListResponse, Error>`
- `RtmClient::im_mark` now returns `Result<api::im::MarkResponse, Error>`

All of them used to return `Result<String, Error>` previously.

mthjones commented:
All squashed! I _think_ this should be good to merge now. Let me know if there's anything I missed!

Every function I added has been tested, but I can't guarantee their coverage as I went almost entirely off the Slack API docs which have been pretty spotty. I covered a few alternative cases with tests where I noticed them. Unfortunately with the scope of this, I think it's going to have to be battle-tested a fair amount.

BenTheElder commented:
Awesome, I'll review and merge ASAP.
After the merge I'll write release docs with the breaking changes and
update the crate info, I'll be sure to note those.
If you don't mind I'll add you to the authors list.
On Oct 22, 2015 9:12 PM, "Matt Jones" <notifications@github.com> wrote:

> All squashed! I *think* this should be good to merge now. Let me know if
> there's anything I missed!
>
> Every function I added has been tested, but I can't guarantee their
> coverage as I went almost entirely off the Slack API docs which have been
> pretty spotty. I covered a few alternative cases with tests where I noticed
> them. Unfortunately with the scope of this, I think it's going to have to
> be battle-tested a fair amount.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BenTheElder/slack-rs/pull/17#issuecomment-150399375>.
>

mthjones commented:
I'm cool with that. Thanks!

BenTheElder commented:
Actually, would you add it to Cargo.toml in the PR? That way I get your
preferred email etc.
On Oct 22, 2015 9:36 PM, "Matt Jones" <notifications@github.com> wrote:

> I'm cool with that. Thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/BenTheElder/slack-rs/pull/17#issuecomment-150403373>.
>

BenTheElder commented:
Looks great. I'm going back over it again since it's such a huge commit but It looks ready to merge.

Fantastic work, like you said I'm sure we'll have to add more tests with time, but this has to be up there with the most well tested and complete slack clients now. Hope your project using it is going well!
